adoc/extensions/sycl_khr_includes.adoc

```diff
@@ -0,0 +1,610 @@
+
+Classes, constants, types and functions provided by the headers listed above
+exist within the [code]#::sycl# namespace as with the standard
+[code]#<sycl/sycl.hpp># header.
+
+Note: there is no strict requirement for the content of those headers, i.e. they
+may contain other functions and classes than documented in this extension. In
+fact, some classes, functions or enumerations are documented to be available
+through multiple headers. Nevertheless, implementations are encouraged to keep
+headers contents as small as reasonably possible to provide maximum benefit for
+users.
+
+[[sec:khr-includes-version]]
+=== [code]#<sycl/khr/includes/version># header
+
+This header contains definitions of:
+- [code]#SYCL_KHR_INCLUDES# macro
+- All macro defined by <<sec:preprocessor-definitions-and-macro>>
+- All macro defined by other extensions supported by an implementation
+
+[code]#<sycl/khr/includes/version># header is included by every other header
+documented by this extension, as well as [code]#<sycl/sycl.hpp>#.
+
+There is no guarantee that the aforementioned macro are defined before the
+inclusion point of [code]#<sycl/khr/includes/version>#.
+
+{note}Implementations providing a dedicated SYCL compiler can predefine those
+macro within the compiler. However, the SYCL standard allows for library-only
+implementations and therefore a portable application should not assume that
+any macro definitions are available until [code]#<sycl/khr/includes/version># is
+included.{endnote}
+
+To properly detect in a portable way if this extension is supported by an
+implementation, the following approach should be used:
+
+[source]
+----
+#if __has_include(<sycl/khr/includes/version>)
+// The extension is supported, but the SYCL_KHR_INCLUDES is not
+// guaranteed to be defined at this point for library-only implementations
+
+#include <sycl/khr/includes/version>
+// Now the SYCL_KHR_INCLUDES macro is guaranteed to be defined for all kinds of
+// implementations
+
+// Use the extension
+#include <sycl/khr/includes/vec>
+
+#else
+// Extension is not supported by an implementation
+// Fallback to the core SYCL standard
+#include <sycl/sycl.hpp>
+#endif
+----
+
+[[sec:khr-includes-backend]]
+=== [code]#<sycl/khr/includes/backend># header
+
+This header file contains everything related to SYCL backend APIs:
+
+- [code]#enum class backend#
+- [code]#backend_traits#
+- [code]#get_native#
+- [code]#make_*#
+
+Note that even though functions defined in the header operate with various
+classes like [code]#platform# or [code]#queue#, there is no guarantee that
+the [code]#<sycl/khr/includes/backend># header provides full definitions for
+those classes. In order to write a portable program, additional [code]##include#
+directives must be used in a SYCL program to ensure availability of the
+corresponding classes.
+
+[[sec:khr-includes-runtime]]
+=== [code]#<sycl/khr/includes/runtime># header
+
+This header groups most of classes defined by 4.6 SYCL runtime classes.
```

Comment:
I would like to also highlight this one. The "most" part is very confusing, I suppose. From the compile-time point of view those classes are to small to give any significant impact being together or separate, so initial version had them together to simplify life for users. But I guess confusion doesn't means simplicity :)

You can expect that I will split those classes into separate headers in the first iteration of applying review feedback. Any comments and suggestions are always welcome, of course.

Reply:
I'm beginning to think that we should only provide the split headers, and not provide grouped headers like this.  If we only provide the split headers, we can get real-life experience with them to figure out which groupings are common, and decide later whether we want to provide them.

Note also that if we choose not to provide any grouped headers, it's trivial for a user to create their own.